Assets/_Metro/Scoring/SimpleAgent/agent.py

```diff
@@ -168,12 +169,199 @@ def get_distance(stationPos, game, segment):
+
+
+    def find_closest(self, current_station, stations):
+        closest = None
+        min_distance = float('inf')
+        for station in stations:
+            distance = GeometryUtils.distance_between_points(station.pos, current_station.pos)
+            if distance < min_distance:
+                min_distance = distance
+                closest = station
+        return closest
+
+    def order_stations(self, station_list):
+        if not station_list:
+            return []
+
+        ordered_list = [station_list[0]]  # Start with the first station
+        remaining_stations = station_list[1:]  # Remaining stations to be ordered
+
+        while remaining_stations:
+            current_station = ordered_list[-1]  # Get the last added station
+            next_station = self.find_closest(current_station, remaining_stations)
+            ordered_list.append(next_station)
+            remaining_stations.remove(next_station)
+
+        return ordered_list
+
+    def get_better_paths(self, cost_function, game_state, update_to_game):
+        """
+        Iteratively generates new paths and evaluates their cost using a specified cost function.
+        Updates the planned paths if a better (lower cost) set of paths is found.
+        The process continues for a maximum of 'num_steps' iterations or until no better path is found.
+
+        Args:
+            cost_function (callable): The name of the static method from CostHandler to use for calculating the cost.
+        """
+        # Ensure that the cost function provided is callable
+        if not callable(cost_function):
+            print(f"Error: The provided cost function is not callable.")
+            return
+
+        # Check if the game state has changed and reinitialize paths if needed
+        if not self.init or self.check_for_changes(game_state):
+            self.initialize_records(game_state)
+
+        # Step 1: Use get_paths to generate new paths
+        try:
+            new_paths = self.get_paths()
+        except NotImplementedError:
+            print("Error: get_paths() must be implemented by a subclass.")
+            return
+
+        # Step 2: Calculate the cost of the new paths using the specified cost function
+        new_cost = cost_function(new_paths)
+
+        # Step 3: If the cost is lower than the current self.cost, update planned_paths and self.cost
+        if new_cost < self.cost:
+            print(f"Found a better path with a lower cost: {new_cost} (previous cost: {self.cost})!")
+            self.planned_paths = new_paths
+            self.cost = new_cost
+            # Send the planned paths to the game using WebSocket
+            if update_to_game:
+                for line_index, station_list in enumerate(self.planned_paths):
+                    for insert_index, station in enumerate(station_list):
+                        insert_station(self.ws, line_index, station.id, insert_index)
```

Comment:
"remove_track" might be needed (?) in the case that a station is completely removed from a track and moved to another track.

Reply:
Good catchup. I was wrong by assuming it will update the route directly. I will add remove_station function for agent to update the design properly.

Reply:
Hello, @WeichenLiu. I fixed the bug, but noticed one interesting thing. The current game does not allow user to design a loop. If that is a bug, I'd like to know. But if that is not a bug, I will assume this requirement in the future. 

In addition, may I know your discord ID and thereby, we can discuss some implementation details there. Thanks!